apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnection.cs

```diff
@@ -261,6 +274,104 @@ public SpannerTransaction BeginReadOnlyTransaction(TransactionId transactionId)
         public Task<SpannerTransaction> BeginTransactionAsync(CancellationToken cancellationToken = default) =>
             BeginTransactionImplAsync(s_readWriteTransactionOptions, TransactionMode.ReadWrite, cancellationToken);
 
+        /// <summary>
+        /// Executes a read-write transaction, with retries as necessary.
+        /// The work to perform in each transaction attempt is defined by <paramref name="asyncWork"/>.
+        /// </summary>
+        /// <param name="asyncWork">The work to perform in each transaction attempt.</param>
+        /// <param name="cancellationToken">An optional token for canceling the call.</param>
+        /// <returns>The value returned by <paramref name="asyncWork"/> if the transaction commits successfully.</returns>
+        /// <remarks><paramref name="asyncWork"/> will be fully run whenever the commit on 
+        /// the <see cref="SpannerTransaction"/> that it receives as a parameter aborts. It must be prepared
+        /// to be called more than once.
+        /// A new <see cref="SpannerTransaction"/> will be passed to <paramref name="asyncWork"/>
+        /// each time it is rerun.
+        /// <paramref name="asyncWork"/> doesn't need to handle the lifecycle of the <see cref="SpannerTransaction"/>,
+        /// it will be automatically committed after <paramref name="asyncWork"/> has finished or rollbacked if an 
+        /// <see cref="Exception"/> (other than because the transaction commit aborted) is thrown by <paramref name="asyncWork"/>.</remarks>
+        public async Task<TResult> RunWithRetriableTransactionAsync<TResult>(Func<SpannerTransaction, Task<TResult>> asyncWork, CancellationToken cancellationToken = default)
+        {
+            GaxPreconditions.CheckNotNull(asyncWork, nameof(asyncWork));
+            PooledSession session = null;
+            try
+            {
+                await OpenAsync(cancellationToken).ConfigureAwait(false);
+                return await SpannerRetryHelpers.ExecuteWithRetryAsync(
+                    CommitAttempt,
+                    RetryCommitOnAbortOptions,
+                    Builder.SessionPoolManager.SpannerSettings.Clock ?? SystemClock.Instance,
+                    Builder.SessionPoolManager.SpannerSettings.Scheduler ?? SystemScheduler.Instance,
+                    cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                session?.Dispose();
+            }
+
+            async Task<TResult> CommitAttempt()
```

Comment:
It feels like this could be in SpannerRetryHelpers. It's possible there's too much context here to easily move it there, but it would be nice to make the retry helper class be more helpful :) That can definitely happen in a later PR though - maybe add a TODO for now?